### PR TITLE
bank: list transaction history of the account

### LINKED ISSUE #10: bank: view transactions history
Each user should be able to see it's own transaction history for the operations that are performed in the account.

### PR DESCRIPTION
**Pull request by gadjevb**
Added DAO for the transaction history, now every transaction that is made by the users will be registered in the database and later be presented by the frontend to the user by his request. Thus, every user is now able to see his(and only his) transactions in page like view.

Fixes #10

### CODE REVIEW

File: frontend/app/app.js [hunk 1]
```diff
@@ -15,24 +16,64 @@ app.config(function ($routeProvider, $locationProvider) {
   $locationProvider.html5Mode(true);
 });
 
-app.controller('TransactionsHistoryCtrl', function ($scope, $http) {
-  // an empty list of transactions should be displayed 
-  // before loading of the history
-  $scope.transactions = [];
+app.controller('TransactionsHistoryCtrl', function($scope, $http) {
+  var pageCounter = 0;
```
Comment (mgenov) [inline frontend/app/app.js]: Why it's not used as scope parameter?  This is going to leave as long as the function leaves.

File: frontend/app/app.js [hunk 2]
```diff
@@ -15,24 +16,64 @@ app.config(function ($routeProvider, $locationProvider) {
...
+  $scope.status = {};
```
Comment (mgenov) [inline frontend/app/app.js]:
```javascript
$scope.status = {next: false, back: false}; 
$scope.query = {id: '', username: 'A', isNext: false}; // What is the meaning of 'A' ?
$scope.records = []
```
Keep in mind that the strange thing about the query is the ID and the Username. They are misleading from my point of view and are not displaying correctly the intention.

File: frontend/app/app.spec.js [hunk 3]
```diff
@@ -40,7 +40,54 @@ describe('TransactionsHistoryCtrl', function () {
...
+     $scope.goNext();
+
+     $httpBackend.flush();
+
+     expect($scope.records).toEqual([{name: 'A' ,id: '34dfs'}, {name: 'A' ,id: '34dfs'}]);
+     expect($scope.query.id).toEqual('25dfs');
+     expect($scope.status.back).toEqual(false);
+  }));
+
+  it('Should go back to previous page', inject(function ($http) {
```
Comment (mgenov) [inline frontend/app/app.spec.js]: You are not handling the cases when there is no next or previous page.

File: src/main/java/com/clouway/bank/adapter/http/OperationsService.java [hunk 4]
```diff
@@ -46,17 +47,32 @@ public OperationsService(AccountRepository accountRepository, UserSecurity userS
...
+        transactionRepository.registerTransaction(new Transaction(
+                null,
+                Date.from(Instant.now()),
+                "A",
+                operation.type,
+                Double.valueOf(operation.amount)
+        ));
+
+        newBalance = account.balance - Double.valueOf(operation.amount);
+        accountRepository.update(account.id, newBalance);
```
Comment (mgenov) [inline src/main/java/com/clouway/bank/adapter/http/OperationsService.java]: Doesn't look nice. The AccountRepository should be responsible for this or some of it's collaborators. 

Service calls Repository and it should be responsible for the creation of History records from the transactions.

File: src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java [hunk 5]
```diff
@@ -1,39 +1,67 @@
...
+    ).as(Json.class);
 
-    return Reply.with(transactionList).as(Json.class);
   }
 
-  private List<Transaction> generateList() {
-    List<Transaction> transactions = new ArrayList<>();
+  @Post
+  public Reply<?> retrieveTransactionHistory(Request request) {
+    System.out.println("Post: ");
```
Comment (mgenov) [inline src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java]: .

File: src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java [hunk 6]
```diff
@@ -1,39 +1,67 @@
...
+    Query query = request.read(Query.class).as(Json.class);
 
-    for (int i = 0; i < 2; i++) {
-      transactions.add(new Transaction(Date.from(LocalDate.of(2017, 1, 1).atStartOfDay(ZoneId.systemDefault()).toInstant()), "testUser", "Deposit", 100d + i));
-    }
+    System.out.println(query.toString());
```
Comment (mgenov) [inline src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java]: .

File: src/test/java/com/clouway/bank/adapter/persistence/PersistentTransactionRepositoryTest.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  public void retrieveFirstPage() {
+    pretendThatExistingTransactionsAre(generateTransactions(27));
+
+    List<Record> firstPage = transactionRepository.retrieveTransactions("A", "", true, 20);
+
+    assertThat(firstPage.size(), is(equalTo(20)));
+  }
+
+  @Test
+  public void retrieveSecondPage() {
```
Comment (mgenov) [inline src/test/java/com/clouway/bank/adapter/persistence/PersistentTransactionRepositoryTest.java]: The scenario of: noNextPage is missing. 

What is the magical thing of "A" ? I saw it in the UI and this looks really strange.

File: src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java (hunk 5 above)
Comment (gadjevb) [inline src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java]: Will be removed!

File: src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java (hunk 6 above)
Comment (gadjevb) [inline src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java]: Will be removed!

**Comment by mgenov** [comment]
I've changed the title of PR to be aligned with the functionality, please make sure that the comment is updated too.

**Comment by gadjevb** [comment]
@mgenov made the required changes, didn't like the way I did the recording of the transactions in the PersistentAcountRepository, but nothing else come to mind!

File: frontend/app/app.spec.js [hunk 8]
```diff
@@ -40,9 +40,85 @@ describe('TransactionsHistoryCtrl', function () {
...
+     $httpBackend.flush();
+
+     expect($scope.records).toEqual([{cursor: '25dfs'}, {cursor: '25dfs'}]);
+     expect($scope.query.cursor).toEqual('34dfs');
+     expect($scope.status.back).toEqual(true);
+     expect($scope.status.next).toEqual(false);
+     expect($scope.pageCounter).toEqual(0);
+  }));
+
+  it('No next page', inject(function ($http) {
```
Comment (mgenov) [inline frontend/app/app.spec.js]: You have to follow the convention. 

it('Should not go forward when there is no next page'....

File: pom.xml [hunk 9]
```diff
@@ -115,7 +115,7 @@
         <groupId>org.codehaus.mojo</groupId>
         <executions>
           <execution>
-            <id>Bundle Script</id>
+            <cursor>Bundle Script</cursor>
```
Comment (mgenov) [inline pom.xml]: this change should be reverted.

File: src/main/java/com/clouway/bank/BankBootstrap.java [hunk 10]
```diff
@@ -60,12 +61,13 @@ public static void main(String[] args) {
...
 
     }
-    if (!successfullyConnnected) {
+    if (!successfullyConnected) {
       System.out.println("unable to connect to the database.");
       System.exit(-1);
     }
 
-    PersistentAccountRepository accountRepository = new PersistentAccountRepository(Providers.of(client.getDatabase("bankApp")));
+    PersistentTransactionRepository transactionRepository = new PersistentTransactionRepository(Providers.of(client.getDatabase("bankApp")));
```
Comment (mgenov) [inline src/main/java/com/clouway/bank/BankBootstrap.java]: client.getDatabase call could be extracted as reference variable as it's duplicated.

File: src/main/java/com/clouway/bank/core/TransactionRepository.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+   * @param id used to match the record from which to begin with
+   * @param isNext used to specify if the next or previous page should
+   *               be rendered to the user. If True the next page will
+   *               be rendered, if False the previous page will be
+   *               rendered
+   * @param limit used to specify the maximum number of transactions
+   *              that could be returned per page
+   * @return List of Record objects that are result of the search
+   */
+  List<Record> retrieveTransactions(String username, String id, Boolean isNext, Integer limit);
```
Comment (mgenov) [inline src/main/java/com/clouway/bank/core/TransactionRepository.java]: String id still remains.

File: src/main/java/com/clouway/bank/adapter/persistence/PersistentTransactionRepository.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    document.put("date", transaction.date);
+    document.put("username", transaction.username);
+    document.put("type", transaction.type);
+    document.put("amount", transaction.amount);
+
+    transactions().insertOne(document);
+  }
+
+  @Override
+  public List<Record> retrieveTransactions(String username, String cursor, Boolean isNext, Integer limit) {
```
Comment (mgenov) [inline src/main/java/com/clouway/bank/adapter/persistence/PersistentTransactionRepository.java]: String cursor could be changed to: String startingFromCursor;

File: src/main/java/com/clouway/bank/core/Record.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
+package com.clouway.bank.core;
+
+/**
+ * @author Borislav Gadjev <gadjevb@gmail.com>
+ */
+public class Record {
```
Comment (mgenov) [inline src/main/java/com/clouway/bank/core/Record.java]: TransactionRecord as it has amount field which is concrete for the transactions.

File: src/test/java/com/clouway/bank/adapter/persistence/PersistentTransactionRepositoryTest.java (hunk 7 above)
Comment (mgenov) [inline src/test/java/com/clouway/bank/adapter/persistence/PersistentTransactionRepositoryTest.java]: .

File: src/main/java/com/clouway/bank/adapter/http/OperationsService.java [hunk 14]
```diff
@@ -1,16 +1,15 @@
...
-import com.clouway.bank.core.UserSecurity;
+import com.clouway.bank.core.*;
 import com.google.inject.Inject;
 import com.google.sitebricks.At;
 import com.google.sitebricks.headless.Reply;
 import com.google.sitebricks.headless.Request;
 import com.google.sitebricks.headless.Service;
 import com.google.sitebricks.http.Post;
 
+import java.sql.Date;
```
Comment (mgenov) [inline src/main/java/com/clouway/bank/adapter/http/OperationsService.java]: java.sql.Date ?

File: src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java [hunk 15]
```diff
@@ -1,39 +1,64 @@
...
+  @Inject
+  public TransactionHistoryService(TransactionRepository transactionRepository) {
+    this.transactionRepository = transactionRepository;
+  }
 
   @Get
   public Reply<?> retrieveTransactionHistory() {
-    transactionList = generateList(); //todo remove the unnecessary generateList method and implement DAO from which the list will be retrieved
+    return Reply.with(
+            transactionRepository.retrieveTransactions("A", "", false, limit)
```
Comment (mgenov) [inline src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java]: You have to retrieve transactions of the current user, not of "A". All these "A" need to be removed I suppose. 

If you are using accounts in tests, you can write: "::any user id::" which is much better then "A". The text will tell that user id is not relevant.

Comment (gadjevb) [inline src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java]: For the moment they are hardcoded, they will be removed at later stage, when there's a way to get the real username.

**Comment by gadjevb** [comment]
@mgenov made the changes according to the comments, will go through the code to check if I missed something. Ready for review!

File: src/main/resources/com/clouway/bank/MainPage.html [hunk 16]
```diff
@@ -16,7 +16,7 @@
             </div>
             <ul class="nav navbar-nav navbar-right">
                 <li><a ng-href="/">HOME</a></li>
-                <li><a ng-href="history">TRANSACTION HISTORY</a></li>
+                <li><a ng-href="v1/transactions">TRANSACTION HISTORY</a></li>
```
Comment (mgenov) [inline src/main/resources/com/clouway/bank/MainPage.html]: Seems broken url rename.

File: src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java [hunk 17]
```diff
@@ -1,39 +1,71 @@
...
+    return Reply.with(
+            transactionRepository.retrieveTransactions(account.userId, "", false, limit)
+    ).as(Json.class);
 
-    return Reply.with(transactionList).as(Json.class);
   }
 
-  private List<Transaction> generateList() {
-    List<Transaction> transactions = new ArrayList<>();
+  @Post
```
Comment (mgenov) [inline src/main/java/com/clouway/bank/adapter/http/TransactionHistoryService.java]: Just saw that you are using post for retrieval which is breaking the REST principles. 

Only GET need to be used: 
/v1/transactions?query=next&abv=false

**Comment by gadjevb** [comment]
@mgenov removed the POST requests, now everything is done through GET request. Ready for review!

**Pull request merged**